contracts/Redemption.vy

```diff
@@ -0,0 +1,489 @@
+X4: constant(int256)  = X3 / 2
+A4: constant(int256)  = 298_095_798_704_172_827_474 * E3
+X5: constant(int256)  = X4 / 2
+A5: constant(int256)  = 5_459_815_003_314_423_907_810
+X6: constant(int256)  = X5 / 2
+A6: constant(int256)  = 738_905_609_893_065_022_723
+X7: constant(int256)  = X6 / 2
+A7: constant(int256)  = 271_828_182_845_904_523_536
+X8: constant(int256)  = X7 / 2
+A8: constant(int256)  = 164_872_127_070_012_814_685
+X9: constant(int256)  = X8 / 2
+A9: constant(int256)  = 128_402_541_668_774_148_407
+X10: constant(int256) = X9 / 2
+A10: constant(int256) = 11_331_4845_306_682_631_683
+X11: constant(int256) = X10 / 2
+A11: constant(int256) = 1_064_49_445_891_785_942_956
+
+@external
+def __init__(
+    yfi: address, d_yfi: address, ve_yfi: address, owner: address, 
+    price_feed: address, curve_pool: address, scaling_factor: uint256,
+):
+    YFI = ERC20(yfi)
+    DYFI = ERC20(d_yfi)
+    VEYFI = ERC20(ve_yfi)
+    PRICE_FEED = AggregatorV3Interface(price_feed)
+    CURVE_POOL = CurvePoolInterface(curve_pool)
+    self._transfer_ownership(owner)
+    self.payee = owner
+    self.packed_scaling_factor = shift(scaling_factor, 128) | shift(scaling_factor, 192)
+
+
+@payable
+@external
+def redeem(amount: uint256, recipient: address = msg.sender) -> uint256:
+    """
+    @notice Redeem your dYFI for YFI using ETH.
+    @dev Redemption tolerates a 0.3% negative or positive slippage.
+    @param amount amount of dYFI to spend
+    @param recipient of the exercised YFI
+    """
+    self._check_killed()
+    assert YFI.balanceOf(self) >= amount, "not enough YFI"
+    eth_required: uint256 = self._eth_required(amount)
+    tolerance: uint256 = eth_required * SLIPPAGE_TOLERANCE / SLIPPAGE_DENOMINATOR
+    if msg.value < (eth_required - tolerance) or msg.value > (eth_required + tolerance):
+        raise "price out of tolerance"
+    IDYFI(DYFI.address).burn(msg.sender, amount)
+    send(self.payee, msg.value)
+    YFI.transfer(recipient, amount)
+    return amount
+
+
+@external
+@view
+def discount() -> uint256:
+    """
+    @notice Get the current dYFI redemption discount
+    @return Redemption discount (18 decimals)
+    """
+    return self._discount()
+
+
+@internal
+@view
+def _discount() -> uint256:
+    yfi_supply: uint256 = YFI.totalSupply()
+    veyfi_supply: uint256 = VEYFI.totalSupply()
+    x: int256 = convert(veyfi_supply * UNIT / yfi_supply, int256)
+    x = self._exp(47 * (self._scaling_factor()[0] * x / E18 - E18) / 10)
+    return convert(E18 * E18 / (E18 + 10 * x), uint256)
+
+
+@external
+@view
+def eth_required(amount: uint256) -> uint256:
+    """
+    @notice Estimate the required amount of ETH to redeem the amount of dYFI for YFI
+    @param amount Amount of dYFI
+    @return Amount of ETH required
+    """
+    return self._eth_required(amount)
+
+
+@internal
+@view
+def _eth_required(amount: uint256) -> uint256:
+    eth_per_yfi: uint256 = amount * self._get_latest_price() / UNIT
```

Comment:
would rename this variable since its the total eth for all the yfi so eth_per_yfi is misleading.